e2e/LongHaul/amqpPooling/IotHub.cs

```diff
@@ -0,0 +1,192 @@
+{
+    internal class IotHub : IAsyncDisposable
+    {
+        private readonly string _hubConnectionString;
+        private readonly Logger _logger;
+        private readonly IotHubClientOptions _clientOptions;
+        private static IList<Device> s_devices;
+        private static IDictionary<string, IotHubDeviceClient> s_deviceClients;
+
+        private static readonly TimeSpan s_messageLoopSleepTime = TimeSpan.FromSeconds(3);
+        private readonly ConcurrentQueue<SystemHealthTelemetry> _messagesToSend = new();
+
+        private SemaphoreSlim _lifetimeControl = new(1, 1);
+
+        private long _totalTelemetryMessagesSent = 0;
+
+        public IotHub(Logger logger, Parameters parameters, IList<Device> devices)
+        {
+            _logger = logger;
+            _hubConnectionString = parameters.IotHubConnectionString;
+            _clientOptions = new IotHubClientOptions(parameters.GetTransportSettingsWithPooling());
+
+            s_devices = devices;
+            s_deviceClients = new Dictionary<string, IotHubDeviceClient>();
+        }
+
+        public Dictionary<string, string> TelemetryUserProperties { get; } = new();
+
+        public async Task InitializeAsync()
+        {
+            await _lifetimeControl.WaitAsync().ConfigureAwait(false);
+
+            var helper = new IotHubConnectionStringHelper(_hubConnectionString);
+
+            try
+            {
+                foreach (var device in s_devices)
+                {
+                    string deviceConnectionString = $"HostName={helper.HostName};DeviceId={device.Id};SharedAccessKey={device.Authentication.SymmetricKey.PrimaryKey}";
+
+                    var deviceClient = new IotHubDeviceClient(deviceConnectionString, _clientOptions);
+
+                    // Set the device client connection status change handler
+                    var amqpConnectionStatusChange = new AmqpConnectionStatusChange(_logger.Clone());
+                    deviceClient.ConnectionStatusChangeCallback = amqpConnectionStatusChange.ConnectionStatusChangeHandler;
+
+                    await deviceClient.OpenAsync().ConfigureAwait(false);
+
+                    s_deviceClients.Add(device.Id, deviceClient);
+                }
+            }
+            finally
+            {
+                _lifetimeControl.Release();
+            }
+        }
+
+        public async Task SendTelemetryMessagesAsync(CancellationToken ct)
+        {
+            var sw = new Stopwatch();
+
+            while (!ct.IsCancellationRequested)
+            {
+                await Task.Delay(s_messageLoopSleepTime, ct).ConfigureAwait(false);
+
+                _messagesToSend.TryDequeue(out SystemHealthTelemetry systemHealth);
+                if (systemHealth == null)
+                {
+                    continue;
+                }
+
+                if (s_deviceClients != null)
+                {
+                    foreach (KeyValuePair<string, IotHubDeviceClient> entry in s_deviceClients)
```

Comment:
I'd rather see independent tasks running for each device, sending telemetry as each one, rather than a loop for each device. This way the telemetry sends can happen at the same time.

Reply:
Consider the pattern we use in the service long haul client where for each device we kick off a Task and then `WhenAll` them.

Reply:
Now I have a `List<Task>` for sending telemetry messages from each device and kick off them to [start ](https://github.com/Azure/azure-iot-sdk-csharp/blob/eea3f874c82f893562abf16e9b206fa2ea5ccf86/e2e/LongHaul/amqpPooling/IotHub.cs#L114)together with `Task.WhenAll`.